### ISSUE TITLE
Order-dependent type error

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#3585]
Consider the following code:

```python
from dataclasses import dataclass
from typing import Protocol

class P1(Protocol):
    x: int

class P2(Protocol):
    y: P1

@dataclass
class A:
    x: int

@dataclass
class B:
    y: A

y: P1 = A(3)
x: P2 = B(A(3))
```

No error is reported, but if the last two lines are swapped, pyright reports an error.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.255, which I just published. It will also be included in the next release of pylance.
